specification/dartLangSpec.tex

```diff
@@ -11677,113 +12041,209 @@ \subsection{Identifier Reference}
+and the interface of the enclosing class has a member named \id.
+Both the static analysis and evaluation proceeds with
+\code{\THIS.\id},
+so there is no need to further specify the treatment of $e$.%
+}
+\EndCase
+
+\LMHash{}%
+Evaluation of an identifier expression $e$ of the form \id{}
+proceeds as follows:
+
+\LMHash{}%
+\Case{Lexical lookup yields a declaration}
+In this case the lexical lookup
+(\ref{lexicalLookup})
+for \id{} yields a declaration $D$.
+The evaluation of $e$ proceeds as follows:
 
 \begin{itemize}
-\item If $d$ is a class, type alias or type parameter the static type of $e$ is \code{Type}.
-\item If $d$ is a local variable (\commentary{which can be a formal parameter})
-  the static type of $e$ is the type of the variable \id,
-  unless \id{} is known to have some type $T$,
-  in which case the static type of $e$ is $T$,
-  provided that $T$ is a subtype of any other type $S$ such that $v$ is known to have type $S$.
-\item If $d$ is a static method, top-level function or local function the static type of $e$ is the function type defined by $d$.
-\item If $d$ is the declaration of a class variable, static getter or static setter declared in class $C$,
-  the static type of $e$ is the static type of the getter invocation (\ref{propertyExtraction}) \code{$C$.\id}.
-\item If $d$ is the declaration of a library variable, top-level getter or top-level setter,
-  the static type of $e$ is the static type of the top level getter invocation \id.
-\item Otherwise, if $e$ occurs inside a top level or static function (be it function, method, getter, or setter) or variable initializer,
-  the static type of $e$ is \DYNAMIC{}.
-\item Otherwise, the static type of $e$ is the type of the property extraction (\ref{propertyExtraction}) \code{\THIS.\id}.
+\item
+  If $D$ is a class, mixin, or type alias,
+  the value of $e$ is an object implementing the class \code{Type}
+  which reifies the corresponding type.
+\item
+  If $D$ is a type parameter $X$ then the value of $e$ is
+  the value of the actual type argument corresponding to $X$
+  that was passed to the generative constructor that created
+  the current binding of \THIS{}.
+\item
+  If $D$ is the declaration of
+  a library variable, top-level getter, or top-level setter,
+  then evaluation of $e$ is equivalent to evaluation of an invocation of
+  the top level getter \id{}
+  (\ref{topLevelGetterInvocation}).
+\item
+  If $D$ is a constant variable of one of the forms
+  \code{\CONST{} $v$ = $e'$;} or \code{\CONST{} $T$ $v$ = $e'$;}
+  then the value of $e$ is the value of the constant expression $e'$.
+\item
+  If $D$ is a declaration of
+  a top-level function, static method, or local function,
+  then $e$ evaluates to the function object obtained by closurization
+  (\ref{functionClosurization})
+  of $D$.
+\item
+  If $D$ is a local variable (\commentary{which can be a formal parameter})
+  then $e$ evaluates to the current binding of \id.
```

Comment:
> ... current binding of that variable. 

(There might be more than one \id in the program, the lookup distinguishes them, so use the result of the lookup. The name itself is underspecified unless you reintroduces scopes).

Reply:
Now using a named variable $v$, such that it is implied that we know more than the name \id. I believe this makes it reasonable to say that it determines the location where its binding (value) can be looked up, corresponding to the scope where we found it at compile-time.